akka-actor-typed/src/main/scala/akka/actor/typed/internal/delivery/ProducerController.scala

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.internal.delivery
+
+import scala.concurrent.duration._
+import scala.reflect.ClassTag
+
+import akka.Done
+import akka.actor.typed.ActorRef
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.ActorContext
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.TimerScheduler
+import akka.actor.typed.scaladsl.LoggerOps
+
+// FIXME Scaladoc describes how it works, internally. Rewrite for end user and keep internals as impl notes.
+
+/**
+ * The `ProducerController` sends `RequestNext` to the actual producer, which is then allowed to send
+ * one message to the `ProducerController`. The `RequestMessage` message is defined via a factory
+ * function so that the producer can decide what type to use. The producer and `ProducerController`
+ * are supposed to be local so that these messages are fast and not lost.
```

Comment:
Interesting pushing the adaption to the sending side instead of requiring a message adapter, is it worth the convenience with another pattern for it though? And if it is, should we use it in more places?

Reply:
I'll take a look to see if we can use a message adapter in the P instead. That would require that P is started, creates the adapter, and then sends an initial Start message to the PC.

On the CC side I used message adapter and such start message.

Reply:
I have completed this refactoring. Now the producer initiates the flow by sending a Start to the ProducerController. I think that is better.